### PR TITLE
MH-11918: AWS S3 Asset Storage

### PR DESCRIPTION
**Pull request by gregorydlogan**
This pull request implements S3 storage for use within the Asset Manager.  This pull request depends on #319.

Major changes:

 * Added support for S3 data storage for the Asset Manager.  File are transparently downloaded to the workspace as needed, rather than requiring restore to local storage prior to usage.  This makes this storage layer _warm_ storage - and thus potentially costly to test!

Known issues:
 * There are no UIs built for this yet. This is strictly back-end functionality

Review suggestion 1: Wait for #319 to be merged.  I will try and keep on top of merging that branch into this one, but it's possible I'll forget.
Review suggestion 2: If you can't wait, there's really only a single commit in the PR (so far), so figure out where that is and review just that commit.

*Work sponsored by the University of Stuttgart*

### CODE REVIEW

**Comment by gregorydlogan** [comment]
I just realized there are a few commits missing here, so I'm putting this on hold until I can get them in.

**Comment by gregorydlogan** [comment]
That's what I get for keeping parts of things in multiple places.  This should be ready to review after #319 gets merged.

File: modules/asset-manager-storage-aws/src/main/java/org/opencastproject/assetmanager/aws/AwsAbstractArchive.java
Comment (rute-santos) [inline modules/asset-manager-storage-aws/src/main/java/org/opencastproject/assetmanager/aws/AwsAbstractArchive.java]: "episode" looks like a 1.x path :)

File: modules/asset-manager-storage-aws/src/main/java/org/opencastproject/assetmanager/aws/AwsAbstractArchive.java
Comment (rute-santos) [inline modules/asset-manager-storage-aws/src/main/java/org/opencastproject/assetmanager/aws/AwsAbstractArchive.java]: Remove part of comment that refers to the dce-specific hack :)

File: docs/upgrade/5_to_6/mysql5.sql.orig
Comment (rute-santos) [inline docs/upgrade/5_to_6/mysql5.sql.orig]: Conflict?

File: etc/workflows/ng-offload.xml
Comment (rute-santos) [inline etc/workflows/ng-offload.xml]: Use 'archive' instead.

File: etc/workflows/ng-restore.xml
Comment (rute-santos) [inline etc/workflows/ng-restore.xml]: Use 'archive' instead.

File: modules/asset-manager-storage-aws/src/main/java/org/opencastproject/assetmanager/aws/s3/AwsS3AssetStore.java
Comment (rute-santos) [inline modules/asset-manager-storage-aws/src/main/java/org/opencastproject/assetmanager/aws/s3/AwsS3AssetStore.java]: Can you please make access key and secret access key optional (as in the s3 distribution service)? We use instance profiles credentials so we don't use keys.

**Review by rute-santos — COMMENTED** [review]
Hi @gregorydlogan ,
Just a few minor issues. I will test everything when I get back from vacation!

File: modules/asset-manager-storage-aws/src/main/java/org/opencastproject/assetmanager/aws/AwsAbstractArchive.java
Comment (gregorydlogan) [reply modules/asset-manager-storage-aws/src/main/java/org/opencastproject/assetmanager/aws/AwsAbstractArchive.java]: If I remember correctly this is actually part of the upstream codebase now, but I'll reword the comment.

File: modules/asset-manager-storage-aws/src/main/java/org/opencastproject/assetmanager/aws/s3/AwsS3AssetStore.java
Comment (gregorydlogan) [reply modules/asset-manager-storage-aws/src/main/java/org/opencastproject/assetmanager/aws/s3/AwsS3AssetStore.java]: Will do!

**Comment by staubesv** [comment]
Hi @gregorydlogan Please note that https://github.com/opencast/opencast/pull/314 has caused conflicts in this PR.

**Comment by gregorydlogan** [comment]
Fixed, and formatting rules applied.

File: docs/guides/admin/docs/modules/awss3archive.md
Comment (rute-santos) [inline docs/guides/admin/docs/modules/awss3archive.md]: This table is not being formatted.  Maybe because the line with
|--------|-----|-------|
is missing?

File: docs/guides/user/docs/advanced/offload-s3.md
Comment (rute-santos) [inline docs/guides/user/docs/advanced/offload-s3.md]: The images are showing up as broken links to me. I did "mkdocs build" and opened it in a browser.

File: docs/guides/admin/docs/modules/awss3archive.md
Comment (rute-santos) [inline docs/guides/admin/docs/modules/awss3archive.md]: Missing woh documentation. Are the woh's included in this PR?

File: docs/guides/admin/docs/modules/awss3archive.md
Comment (rute-santos) [inline docs/guides/admin/docs/modules/awss3archive.md]: Broken link.

File: etc/workflows/ng-offload.xml
Comment (rute-santos) [inline etc/workflows/ng-offload.xml]: Deprecated prefix: "ng-".

File: etc/workflows/ng-restore.xml
Comment (rute-santos) [inline etc/workflows/ng-restore.xml]: Deprecated prefix: "ng-".

File: etc/org.opencastproject.assetmanager.aws.s3.AwsS3AssetStore.cfg
Comment (rute-santos) [inline etc/org.opencastproject.assetmanager.aws.s3.AwsS3AssetStore.cfg]: Comment needs to be fixed. There's no default value.

Comment (rute-santos) [reply etc/org.opencastproject.assetmanager.aws.s3.AwsS3AssetStore.cfg]: I am wondering if it would be best to have a "enabled=true/false" property because an error will show up in the log if this service is not being used: 
2018-09-07 13:36:17,470 | ERROR | (?:?) - [org.opencastproject.assetmanager.aws.s3.AwsS3AssetStore(212)] The activate method has thrown an exception
org.opencastproject.util.ConfigurationException: org.opencastproject.assetmanager.aws.s3.bucket is missing or invalid

File: docs/guides/admin/docs/modules/awss3archive.md
Comment (gregorydlogan) [reply docs/guides/admin/docs/modules/awss3archive.md]: Ah, so they are but there's no way to reliably link between the documentation sets (ie, the docs are in user, can't link across from admin).  I'll rework this.

File: docs/guides/user/docs/advanced/offload-s3.md
Comment (gregorydlogan) [reply docs/guides/user/docs/advanced/offload-s3.md]: Images went missing in the numerous rebases.  Good thing I still have the base codebase!

File: docs/guides/admin/docs/modules/awss3archive.md
Comment (gregorydlogan) [reply docs/guides/admin/docs/modules/awss3archive.md]: The header and section formatting also does not match the rest of the docs since these were written prior the recent style changes.  I'll fix that while I'm at it.

File: docs/guides/admin/docs/modules/awss3archive.md
Comment (gregorydlogan) [reply docs/guides/admin/docs/modules/awss3archive.md]: Broken link and completely missing documentation...

File: etc/org.opencastproject.assetmanager.aws.s3.AwsS3AssetStore.cfg
Comment (gregorydlogan) [reply etc/org.opencastproject.assetmanager.aws.s3.AwsS3AssetStore.cfg]: That's fair, done.

**Comment by gregorydlogan** [comment]
@rute-santos, ready for review again I think.

File: etc/workflows/offload.xml
Comment (rute-santos) [inline etc/workflows/offload.xml]: The id still has "ng-". Wouldn't something like offload-to-remote/restore-from-remote or something similar be better names? Offload is pretty clear, but 'restore' is ambiguous imho. But whatever you think is best.

File: etc/workflows/restore.xml
Comment (rute-santos) [inline etc/workflows/restore.xml]: The id still has "ng-"

File: docs/guides/admin/docs/modules/awss3archive.md
Comment (rute-santos) [inline docs/guides/admin/docs/modules/awss3archive.md:33]: Missing the org.opencastproject.assetmanager.aws.s3.enabled configuration.

File: docs/guides/admin/docs/modules/awss3archive.md
Comment (rute-santos) [inline docs/guides/admin/docs/modules/awss3archive.md]: Please remove "ng-".

File: docs/guides/admin/docs/workflowoperationhandlers/move-storage-woh.md
Comment (rute-santos) [inline docs/guides/admin/docs/workflowoperationhandlers/move-storage-woh.md]: This example doesn't look right...

**Comment by KatrinIhler** [comment]
@gregorydlogan Conflicts again...

**Comment by gregorydlogan** [comment]
@KatrinIhler ty!

@rute-santos fixes applied, ready for further review.

**Review by rute-santos — APPROVED** [review]

**Pull request merged**
